mods/stairs/init.lua

```diff
@@ -129,86 +134,58 @@ function stairs.register_slab(subname, recipeitem, groups, images, description,
-			local p0 = pointed_thing.under
-			local p1 = pointed_thing.above
-			local n0 = minetest.get_node(p0)
-			local n1 = minetest.get_node(p1)
-			local param2 = 0
+			local dir = minetest.dir_to_wallmounted(vector.subtract(
+				pointed_thing.above, pointed_thing.under), true)
 
-			local n0_is_upside_down = (n0.name == "stairs:slab_" .. subname and
-					n0.param2 >= 20)
-
-			if n0.name == "stairs:slab_" .. subname and not n0_is_upside_down and
-					p0.y + 1 == p1.y then
-				slabpos = p0
-				slabnode = n0
-			elseif n1.name == "stairs:slab_" .. subname then
-				slabpos = p1
-				slabnode = n1
+			local rot = slab_trans_dir_place[dir]
+			if rot == 0 or rot == 20 then
+				rot = rot + minetest.dir_to_facedir(placer:get_look_dir())
 			end
-			if slabpos then
-				-- Remove the slab at slabpos
-				minetest.remove_node(slabpos)
-				-- Make a fake stack of a single item and try to place it
-				local fakestack = ItemStack(recipeitem)
-				fakestack:set_count(itemstack:get_count())
-
-				pointed_thing.above = slabpos
-				local success
-				fakestack, success = minetest.item_place(fakestack, placer,
-					pointed_thing)
-				-- If the item was taken from the fake stack, decrement original
-				if success then
-					itemstack:set_count(fakestack:get_count())
-				-- Else put old node back
-				else
-					minetest.set_node(slabpos, slabnode)
-				end
+
+			-- place_node will in turn attempt to call on_rightclick()
+			return minetest.item_place(itemstack, placer, pointed_thing, rot)
+		end,
+		on_rightclick = function(pos, node, clicker, itemstack, pointed_thing)
+			if not pointed_thing then
 				return itemstack
 			end
-			
-			-- Upside down slabs
-			if p0.y - 1 == p1.y then
-				-- Turn into full block if pointing at a existing slab
-				if n0_is_upside_down  then
-					-- Remove the slab at the position of the slab
-					minetest.remove_node(p0)
-					-- Make a fake stack of a single item and try to place it
-					local fakestack = ItemStack(recipeitem)
-					fakestack:set_count(itemstack:get_count())
-
-					pointed_thing.above = p0
-					local success
-					fakestack, success = minetest.item_place(fakestack, placer,
-						pointed_thing)
-					-- If the item was taken from the fake stack, decrement original
-					if success then
-						itemstack:set_count(fakestack:get_count())
-					-- Else put old node back
-					else
-						minetest.set_node(p0, n0)
+			local dir = minetest.dir_to_facedir(vector.subtract(
+				pointed_thing.above, pointed_thing.under), true)
+
+			local wield_item = clicker:get_wielded_item():get_name()
+			local player_name = clicker:get_player_name()
+			local p2 = node.param2
+
+			if wield_item == node.name then
+				-- combine two slabs
+				if slab_trans_dir[math.floor(p2 / 4)] == dir then
+					-- combinable slab?
+					if not recipeitem then
+						return itemstack
+					end
+					if minetest.is_protected(pos, player_name) and not
+							minetest.check_player_privs(clicker, "protection_bypass") then
```

Comment:
You should use `minetest.get_player_privs(player_name).protection_bypass` instead, if you just check one priv like this.
It is internally called by `check_player_privs`, too, which does some unnecessary set to list conversation, you don't need here.
`check_player_privs` also has a broken signature of return values (set expected and either string or list returned) and should probably be deprecated one day.